### PR TITLE
[Feature] Supports parameterized S3Artifactory for Pipeline and ContainerOp in kfp package

### PR DESCRIPTION
**Pull request by eterna2**
## Motivation
I am running a kubeflow pipeline deployment with my custom helm chart and a minio s3 gateway to my custom bucket. This bucket has a different name from the default one in `kfp`, hence I need some way to parameterize the s3 artifact configs.

## Status
- Waiting for Review

## Features
- kfp can now declare a custom artifact location inside a pipeline or containerop.

```py
from kfp import dsl
from kubernetes.client.models import V1SecretKeySelector


@dsl.pipeline( name='foo', description='hello world')
def foo_pipeline(namespace: str):

    # configures artifact location
    artifact_location = dsl.ArtifactLocation.s3(
                            bucket="foobar",
                            endpoint="minio-service.%s:9000" % namespace,  # parameterized namespace
                            insecure=True,
                            access_key_secret=V1SecretKeySelector(name="minio", key="accesskey"),
                            secret_key_secret={"name": "minio", "key": "secretkey"}  # accepts dict also
    )

    # set pipeline level artifact location
    conf = dsl.get_pipeline_conf().set_artifact_location(artifact_location)
    
    # use pipeline level artifact location (i.e. minio-service)
    op1 = dsl.ContainerOp(name='foo', image='bash:latest')

    # use containerop level artifact location (i.e. aws)
    op2 = dsl.ContainerOp(
                        name='foo', 
                        image='bash:latest',
                        # configures artifact location
                        artifact_location=dsl.ArtifactLocation.s3(
                            bucket="foobar",
                            endpoint="s3.amazonaws.com",
                            insecure=False,
                            access_key_secret=V1SecretKeySelector(name="s3-secret", key="accesskey"),
                            secret_key_secret=V1SecretKeySelector(name="s3-secret", key="secretkey"))
    )

```

## TLDR changes
- `argo-models` is now a dependency in `setup.py` (argo `v2.2.1`)
- Added static class `ArtifactLocation` 
  - to help generate artifact location for s3
  - to help generate artifact for workflow templates
- Updated `PipelineConf` to support artifact location
- Updated k8s helper and related, to support openapi objects (I accidentally used openapi generator instead of swagger codegen for `argo-models`)
- Added unit test for `ArtifactLocation`
- Fixed unit test for `kfp.aws` (Found that it has a bug, and was not imported into the unit test)

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/1064)

### CODE REVIEW

**Comment by hongye-sun** [comment]
@eterna2, I am interested in your use case here. Are you going to export your own data from the container? Or this change is only for metadata and metrics?

Have you checked that argo supports to configure the [default artifact repo](https://github.com/argoproj/argo/blob/master/ARTIFACT_REPO.md#configure-the-default-artifact-repository) in configmap? What happens if we allow user to change the config in KFP cluster? What else is missing from your perspective?

**Comment by Ark-kun** [comment]
> Have you checked that argo supports to configure the [default artifact repo](https://github.com/argoproj/argo/blob/master/ARTIFACT_REPO.md#configure-the-default-artifact-repository) in configmap?

Unfortunately, the hard-coded values in the pipeline.yaml override that config. Hard-coding strikes again...

>What happens if we allow user to change the config in KFP cluster? What else is missing from your perspective?

That would be great.

**Comment by Ark-kun** [comment]
> I need some way to parameterize the s3 artifact configs.

Do you need different configs for different ops? We're gearing more towards having a configurable default storage solution which will be used by all ops (and probably all pipelines).

Maybe we should move this artifact config to the compiler arguments?

    compiler.Compile(pipeline_func, storage_system=ArgoArtifacts(S3Artifactory(...))

Or, like @hongye-sun mentioned, moving the artifact config application to the backend.

What do you think?

**Comment by eterna2** [comment]
@hongye-sun @Ark-kun

> @eterna2, I am interested in your use case here. Are you going to export your own data from the container? Or this change is only for metadata and metrics?

For now only metadata and metrics, but ideally is for all the artifacts in the pipeline (I hope the dsl for persistent volume gets thru). 

I am supporting different teams which may be using the same pipeline. However, I what to separate all artifacts and side effects for different teams into different buckets. So I want to parameterize the artifact config at run time.

> Do you need different configs for different ops We're gearing more towards having a configurable default storage solution which will be used by all ops (and probably all pipelines).
>
> Maybe we should move this artifact config to the compiler arguments?

Realistically, I don't need diff config for diff ops, but depending on use case I might want to do it at pipeline, or containerop level.

For compiler args, I am thinking why not set the fields blank for default (hence we can use the config map for workflow controller).

Pipeline level config is needed just because I want to parameterize them so I can set where to save at run time to support different teams.

ContainerOps is more for running single component ( there are some use cases where we just want to submit a single job ).

I will probably be building some UI over (or customizng) the default kf UI so some of these artifactory settings are not exposed (automatically filled via the individual jwt cookie).

**Comment by Ark-kun** [comment]
> I am thinking why not set the fields blank for default

Yes. I'm in favor of that.
That's what Argo has traditionally been doing. https://github.com/argoproj/argo/blob/master/examples/artifact-passing.yaml

**Comment by eterna2** [comment]
Just to add more context and my wishlist (actually what I wanted to do after this pr):

Currently, I am using kf pipelines to run simulations. But because there is currently no way to declare the output artifacts, the logs and results are separately saved to some bucket by some scripts in the container.

This creates unnecessary complexity as i need to maintain a config to s3 client inside the container. Probably solved by #998. And #801 in the future (for downstream processing of the simulation results which can be very big).

I also hope there is a easier way to declare ui-metadata and integrate with the front-end. Currently, I have to write a scripts in my container to generate these JSON. 

Preferably, I hold I can make it more declarative in containerOp rather than in the light weight python components or inside the container - it mixes the business logic with the operation code.

Something like this, which additional scripts are added to convert files to metadata or metrics or ... 
```py

ContainerOp(..., output_artifacts={'/results/abc.csv': TableView(name='foo',columns=[ ]), '/metrics/f1.txt': Metrics('f1', ...) })
```

**Comment by hongye-sun** [comment]
cc @neuromage 

@eterna2, thanks for sharing your thoughts here. We also have some internal discussion about artifact and metadata. Argo's artifact is great for archiving purpose. However, it's not so great for data passing and realtime debugging/visualization purpose. The way it implements is inefficient for passing large data between steps. We recommend to use volume to pass large data. For data visualization, Argo artifact is only available until the step is finished. In a data processing pipeline, a single step can take a long time to finish. We'd like the visualization to be in realtime, so that user can see the data earlier and decide whether to stop or continue the execution. Another aspect is that the artifacts are not always produced from a file/folder inside a container. It can be produced by external jobs, it can be a S3 file, a HDFS directory, or a SQL table, etc. 

I don't have an answer yet, but with all the above requirements in mind, I am leaning toward letting container code to write the artifact to a remote storage (GCS, S3, NFS, BigQuery table, etc) and report the location and extra metadata to KFP via API or DSL definition. If we go with this approach, we will probably follow what tensorflow filesystem SDK does to abstract the file IO API from the underlying storage layer. 

We might continue support artifact feature as an option if user don't care about realtime visualization and just want to use native file IO API. 

WDYT?

**Comment by eterna2** [comment]
@hongye-sun

I totally agree with the requirements - in fact I needed them yesterday. Especially, a button to kill the job. We are currently looking at the streaming log to see how the simulation is running, and I have to manually kill the pod if it is going south.


```
requirements in mind, I am leaning toward letting container code to write the artifact to a remote storage (GCS, S3, NFS, BigQuery table, etc) and report the location and extra metadata to KFP via API or DSL definition. If we go with this approach, we will probably follow what tensorflow filesystem SDK does to abstract the file IO API from the underlying storage layer.
```

Yup. This is probably a better way. My main concerns were mostly about having 2 diff places to specify where and how to upload the artifacts.

I have some initial thoughts thou. Could we orchestra the io with a configmap volume mount? Mostly because I just want to have a single version of docker image without embedding io codes inside. (Actually, it is to use an existing images without building another one - easier to convince my data scientists to transit to kf).

We probably want to use the kfp sdk to generate this configmap to export the artifacts?

As for streaming vis, this is something we really wanted to. We are doing a combination of geospatial and resource allocation simulation. Originally, I was thinking of just adding a sidecar with either a grpc or just a cron to read from some log output and serve as a proxy. 

Then build a dashboard and point to the sidecar proxy.

**Comment by hongye-sun** [comment]
@eterna2 

The terminate run button is recently added in https://github.com/kubeflow/pipelines/pull/1039. It should be available in recent KFP release.

Could you elaborate a little bit about the configmap volume mount? I don't understand how it works.

Have you considered to have an NFS volume to be mounted in both task pod and UI pod to solve the realtime viz problem? You can also use the same volume to pass data between steps.

**Comment by Ark-kun** [comment]
> Argo's artifact is great for archiving purpose. However, it's not so great for data passing and realtime debugging/visualization purpose. The way it implements is inefficient for passing large data between steps. We recommend to use volume to pass large data.

The important note is that volume-based artifact passing can still be used even when Argo-style artifacts are specified. Argo pipeline can just specify that component1 has an artifact output (data grabbed from local path) and component2 has and artifact input (data placed into local path) and the data should be passed between them.
In the backend, instead of using Argo way of artifact passing implementation, we can actually use volumes to pass the data around. That's without any changes to the pipeline.yaml.

Removing the explicit storage configuration (but not the artifact definitions or local paths) from the shared pipeline.yaml files is the first step to the storage system flexibility. It can be even possible to switch the storage solution (volume-based vs. GCS vs. Minio) right before launching the run.

**Comment by eterna2** [comment]
@hongye-sun 
```
Could you elaborate a little bit about the configmap volume mount? I don't understand how it works.
```
The motivation is to be able to create ContainerOp with existing images (pre-kubeflow and pre-argo era), instead of building a new image with the io codes required to handle the artifacts - I want some way to execute these IO codes without needing to build a new image.

Ok I just figure out I don't need a configmap anymore. I can do it with the exit handler instead.  But does this mean I would need to use a NFS?

```py
import kfp.components as comp
from kfp.dsl import ContainerOp, ExitHandler

def io_handler():
  """custom handling of artifacts"""
  import json
  
  with open("/metrics.json", "w") as f:
    json.dump(f, {})

io_op = comp.func_to_container_op(io_handler)

with ExitHandler(exit_op=io_op):
  op = ContainerOp(image="some_image_wo_io_codes")
```

```
Have you considered to have an NFS volume to be mounted in both task pod and UI pod to solve the realtime viz problem? You can also use the same volume to pass data between steps.
```

Never thought of that. But I think this is a good idea. Might need to try it out - just worried about the write speed, AWS EFS is notoriously slow.

**Comment by eterna2** [comment]
@hongye-sun 
- Resync with current HEAD
- Updated `PipelineConf` and `ContainerOp` to have no `S3Artifactory` configs, i.e. generates pipelines w/o s3 artifacts
- Added `output_artifacts` arg to `ContainerOp` which can be used to overwrite or add custom output artifacts.

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: foo-
spec:
  arguments:
    parameters:
    - name: bucket
      value: foobucket
  entrypoint: foo
  serviceAccountName: pipeline-runner
  templates:
  - dag:
      tasks:
      - name: foo
        template: foo
    name: foo
  - container:
      image: busybox:latest
    name: foo
    outputs:
      artifacts:
      - name: logs
        path: /logs
      - name: mlpipeline-metrics
        path: /foo-bar.json
      - name: mlpipeline-ui-metadata
        path: /mlpipeline-ui-metadata.json
      - name: results
        path: /results.csv
```

**Comment by Ark-kun** [comment]
> The motivation is to be able to create ContainerOp with existing images (pre-kubeflow and pre-argo era), instead of building a new image with the io codes required to handle the artifacts - I want some way to execute these IO codes without needing to build a new image.

I agree that this is an important use case. There are thousands of command-line programs that operate on files. It should be effortless to use them in pipelines.

**Comment by vicaire** [comment]
@Ark-kun, @hongye-sun, do you have any additional comments on this PR?

**Comment by Ark-kun** [comment]
BTW,  Here is my PR which makes the artifact storage configurable by the cluster admin: https://github.com/kubeflow/pipelines/pull/513

**Comment by Ark-kun** [comment]
Naming:

`S3Artifactory` is not a good name, because Artifactory and S3 are different storage services.

`set_s3_artifactory` -> `set_artifact_repository`
`S3Artifactory` -> `S3ArtifactRepository` or `S3ArtifactConfig`

**Comment by hongye-sun** [comment]
@eterna2, sorry for pending this PR for so long. This is something we have thought about but never had a conclusion internally. We had several internal discussions about the PR and the conclusion is that we want to expose Argo's ArtifactLocation type just like you did for Container type for now until we have a better artifact support from our side. 

I like the way that you allow user to add the artifact store template on pipeline or step level. In order to generalize the pattern instead of just s3. Could you change the code like:

```python
   artifact_location = ArtifactLocation.s3(bucket = bucket, endpoint = 's3.amazonaws.com', insecure = False, region='...', access_key_secret = '...', ...)
    get_pipeline_conf().set_artifact_repository(artifact_location)
```

Right now, you may just support S3 location, and later, if some other user want to support git, it can be:
```python
  ArtifactLocation.git(...)
```

**Comment by Ark-kun** [comment]
BTW, @eterna2 Would your original issue be solved if we supported Argo's way of configuring the artifact storage?

Argo has a pretty clean and portable way to use artifacts - in the Workflow you only specify the name and path and the artifact repository configuration is taken from the configMap and applied at runtime by the controller:

https://github.com/argoproj/argo/blob/master/ARTIFACT_REPO.md#configure-the-default-artifact-repository

https://github.com/argoproj/argo/blob/4e37a444bde2a034885d0db35f7b38684505063e/examples/artifact-passing.yaml#L26..L29
```yaml
    outputs:
      artifacts:
      - name: hello-art
        path: /tmp/hello_world.txt
```

If we support this feature (pre-install configMap that you can tweak later and get storage config from it), would this cover your needs?

In Argo 2.4 they'll also add a per-workflow config name option which will make things more flexible:
https://github.com/argoproj/argo/issues/1332

```yaml
kind: Workflow
metadata:
  generateName: output-artifact-s3-
spec:
  entrypoint: whalesay
  archiveLocationConfig:
    name: archive-location-configmap
    key: config
```

**Comment by eterna2** [comment]
@hongye-sun 
```
 artifact_location = ArtifactLocation.s3(bucket = bucket, endpoint = 's3.amazonaws.com', insecure = False, region='...', access_key_secret = '...', ...)
    get_pipeline_conf().set_artifact_repository(artifact_location)
```
Sure can do. I can make a `io.argoproj.workflow.v1alpha1.ArtifactLocation` class which can be used in argo 2.4 as mentioned by @Ark-kun 

I will also probably add a method to `ArtifactLocation` to returns a `Callable[[str, str], io.argoproj.workflow.v1alpha1.Artifact ]`, which can then be used in the compilation step to generate the artifact for argo < 2.4?

I will add in the scripts to generate these argo swagger classes, but I will probably only add the methods for s3 artifacts for now?

Give me a couple of days to resync with the latest branch and remember what I did. lol.

@Ark-kun 
```
BTW, @eterna2 Would your original issue be solved if we supported Argo's way of configuring the artifact storage?
```
Yeah, I've been doing it this way all (and i removed the hardcoded artifact store codes in kfp). 

But I still have an use case for parameterized artifact location. Because unless I am running multiple kubeflow/argo services, different teams have to share the same artifact store.

Also wanted to add that I can't deploy multiple kubeflow services for without doing some proxying magic because the `minio-service.kubeflow:9000` is hardcoded in quite a few places (the frontend server for example).

**Comment by eterna2** [comment]
@hongye-sun @Ark-kun 
Hey, I noticed ur moved the codegen scripts to the backend api as a separate python package.

Do u want me to put the argo models in the kfp package (i.e. add the code gen into the build script), or do I generate a separate python package for the argo models and probably push it to pypi? Cuz I am not sure if there is an existing py package for argo (i searched and did not find any).

**Comment by hongye-sun** [comment]
@eterna2, I didn't find any sdk either. I'd prefer starting with same package as the first step. You may consider to split the PR into smaller PRs which will be easier for us to iterate and review. For example, it can be 3 PRs:

1. Change build script to generate Argo models in kfp's package
2. DSL artifact API to help to build Argo artifact models (only support S3 for this PR)
3. Expose DSL interface to accept artifacts and change compiler to merge it into Argo template

**Comment by Ark-kun** [comment]
> Do u want me to put the argo models in the kfp package (i.e. add the code gen into the build script), or do I generate a separate python package for the argo models and probably push it to pypi?

It would be best to have a separate Argo client package on PyPI. It can be useful for other purposes too.
We want to get rid of `sdk/python/build.sh`.

**Comment by Ark-kun** [comment]
>But I still have an use case for parameterized artifact location. Because unless I am running multiple kubeflow/argo services, different teams have to share the same artifact store.

My Argo PR fixes this issue:
https://github.com/argoproj/argo/pull/1350

You'll be able to specify the configMap name in the Workflow  spec:

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
spec:
  artifactRepositoryRef:
    configMap: artifact-repository
    key: config
...
```

I'll try to backport it to our backend.

**Comment by eterna2** [comment]
@hongye-sun @Ark-kun 
I have created a repo to generate the openapi objects for argo at  https://github.com/e2fyi/argo-models

The pypi package is `argo-models`: https://pypi.org/project/argo-models/

Currently, I provide 2 releases:
- `2.2.1`
- `2.3.0rc3`

I noticed that in `2.3.0rc3`, `v1alpha1_sidecar` was renamed to `v1alpha1_user_container`. Do u prefer me to use `2.3.0` or `2.2.1` for the PR?

**Comment by hongye-sun** [comment]
@eterna2 I think the version should tie to the current argo version we used in KFP's install. I'd say it's 2.2.1. We are working on moving it to 2.3.0 but not completed yet.

**Comment by eterna2** [comment]
@hongye-sun @Ark-kun 

Hey I have removed my previous changes, and reimplemented everything using `argo-models`. 

```py
from kfp import dsl
from kubernetes.client.models import V1EnvVar, V1SecretKeySelector


@dsl.pipeline( name='foo', description='hello world')
def foo_pipeline(namespace: str):

    # configures artifact location
    artifact_location = dsl.ArtifactLocation.s3(
                            bucket="foobar",
                            endpoint="minio-service.%s:9000" % namespace,  # parameterized namespace
                            insecure=True,
                            access_key_secret=V1SecretKeySelector(name="minio", key="accesskey"),
                            secret_key_secret=V1SecretKeySelector(name="minio", key="secretkey"))

    # set pipeline level artifact location
    conf = dsl.get_pipeline_conf().set_artifact_location(artifact_location)
    
    # use pipeline level artifact location (i.e. minio-service)
    op1 = dsl.ContainerOp(name='foo', image='bash:latest')

    # use containerop level artifact location (i.e. aws)
    op2 = dsl.ContainerOp(
                        name='foo', 
                        image='bash:latest',
                        # configures artifact location
                        artifact_location=dsl.ArtifactLocation.s3(
                            bucket="foobar",
                            endpoint="s3.amazonaws.com",
                            insecure=False,
                            access_key_secret=V1SecretKeySelector(name="s3-secret", key="accesskey"),
                            secret_key_secret=V1SecretKeySelector(name="s3-secret", key="secretkey"))
    )

```

File: sdk/python/tests/components/test_python_op.py [hunk 1]
```diff
@@ -228,7 +229,10 @@ def calc_pipeline(
             #Compiling the pipleine:
             pipeline_filename = str(Path(temp_dir_name).joinpath(calc_pipeline.__name__ + '.pipeline.tar.gz'))
             import kfp.compiler as compiler
-            compiler.Compiler().compile(calc_pipeline, pipeline_filename)
+
+            with warnings.catch_warnings():  # ignore pending deprecation warnings
```
Comment (Ark-kun) [inline sdk/python/tests/components/test_python_op.py]: Which deprecated member does this use?

Comment (eterna2) [reply sdk/python/tests/components/test_python_op.py]: If `artifact_location` is not provided, I generates a pending deprecation warning and return the default value for now (e.g. minio-service:9000, ...).

**Comment by Ark-kun** [comment]
Sorry for the merge conflict.
We've just merged the https://github.com/kubeflow/pipelines/pull/1297 which removed the artifact repository configuration from the compiled workflows. The pipelines are now clean and storage-agnostic by default.

    outputs:
      artifacts:
      - name: mlpipeline-ui-metadata
        path: /mlpipeline-ui-metadata.json
        optional: true
      - name: mlpipeline-metrics
        path: /mlpipeline-metrics.json
        optional: true

**Comment by Ark-kun** [comment]
@eterna2 What's the best way to contact you in a [voice] chat-like environment? Hangouts, slack, etc.

**Comment by eterna2** [comment]
@Ark-kun

I am in the kubeflow slack channel under the same nick.

But I am at GMT+8.

File: sdk/python/tests/components/test_python_op.py (hunk 1 above)
Comment (eterna2) [reply sdk/python/tests/components/test_python_op.py]: removed after sync with latest HEAD where default artifact location is empty.

**Comment by eterna2** [comment]
@Ark-kun 
Fixed merge conflicts and removed PendingDeprecation warning.

Empty `artifact_location` will now return an empty artifact config.

File: sdk/python/pipeline.yaml [hunk 2]
```diff
@@ -0,0 +1,170 @@
+apiVersion: argoproj.io/v1alpha1
```
Comment (Ark-kun) [inline sdk/python/pipeline.yaml]: This file is probably not needed here.

File: sdk/python/tests/components/test_python_op.py [hunk 3]
```diff
@@ -228,6 +228,7 @@ def calc_pipeline(
             #Compiling the pipleine:
             pipeline_filename = str(Path(temp_dir_name).joinpath(calc_pipeline.__name__ + '.pipeline.tar.gz'))
             import kfp.compiler as compiler
+
```
Comment (Ark-kun) [inline sdk/python/tests/components/test_python_op.py]: Nit: Stray line?

**Comment by Ark-kun** [comment]
@eterna2 Please remove the stray temporary file and I guess we're done.
Your PR is very nice and clean now.

I still do not think that per-op artifact configuration is a good idea, but I won't hold this PR any more.

I'd like to work with you in future to switch to per-run configuration so that nothing gets hard-coded in the pipeline package.

**Comment by Ark-kun** [comment]
@gaoning777 @hongye-sun Do you see any remaining problems?

File: sdk/python/tests/compiler/testdata/artifact_location.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from kfp import dsl
+from kubernetes.client.models import V1SecretKeySelector
+
+
+@dsl.pipeline(name='foo', description='hello world')
+def foo_pipeline(tag: str, namespace: str = "kubeflow", bucket: str = "foobar"):
```
Comment (hongye-sun) [inline sdk/python/tests/compiler/testdata/artifact_location.py]: Do you mind adding a sample similar with this one in the basic sample folder in another PR? Thanks.

**Comment by Ark-kun** [comment]
/hold

Waiting for https://github.com/kubeflow/pipelines/pull/1331 release to be merged
